respy/shared.py

```diff
@@ -282,15 +276,32 @@ def create_base_covariates(states, covariates_spec, raise_errors=True):
     """
     covariates = states.copy()
 
-    for covariate, definition in covariates_spec.items():
-        if covariate not in states.columns:
-            try:
-                covariates[covariate] = covariates.eval(definition)
-            except pd.core.computation.ops.UndefinedVariableError as e:
-                if raise_errors:
-                    raise e
+    n_covariates_left_changed = True
+    covariates_left = list(covariates_spec)
+
+    while n_covariates_left_changed:
+        n_covariates_left = len(covariates_left)
+
+        # Create a copy of `covariates_left` to remove elements without side-effects.
+        for covariate in copy.copy(covariates_left):
+            # Check if the covariate does not exist and needs to be computed.
+            is_covariate_missing = covariate not in covariates.columns
+
+            if is_covariate_missing:
+                try:
+                    covariates[covariate] = covariates.eval(covariates_spec[covariate])
+                except pd.core.computation.ops.UndefinedVariableError as e:
+                    if raise_errors:
+                        raise e
+                    else:
+                        pass
                 else:
-                    pass
+                    covariates_left.remove(covariate)
+
+        n_covariates_left_changed = n_covariates_left != len(covariates_left)
+
+    if covariates_left and raise_errors:
```

Comment:
You are right. This is a copy-paste-mixup. The if block under the except must be removed and than this is correct.

Yes, we cannot simply always raise errors when there are covariates left. For example, while sampling the experiences in an n-step-ahead, we know there we can use information on observables because they are sampled prior to observables. But other information like previous choices is not available. Separating covariates which rely on previous choices from the rest is more troublesome than necessary.

Reply:
Point well taken!